### PR TITLE
Refactor KeyboardSelect to use Hooks/functional component(s)

### PR DESCRIPTION
**Pull request by pm0u**
## What

- Refactors the KeyboardSelect component to be a functional component (and _nearly_ side effect free)
- Creates some smaller child components with the aim of simplifying the rendering logic of KeyboardSelect
- Moves the business logic of KeyboardSelect to a hook

## Why

- KeyboardSelect was pretty dense (for me) and it was pretty tightly coupled to the hardware. This (along with possible further refactoring) should help move this towards being presentational & more easily debugged & tested
- Along with the first point, separating the interaction with hardware will help to be able to mock dependencies and data and more easily test UI without needing hardware present.

# Note

This is currently a draft, as without a keyboard (atreus someday?), I have _no way_ to test this. I did my best to keep implementation & logic faithful to the original code, however there will more than likely be bugs. heck, this probably will break the moment someone actually tries to detect a keyboard as i can't even test that.

I am not sure if this is a direction that the UI would like to go, it felt to me like this component had grown to be pretty far reaching, and separating some things out while moving towards the use of hooks and functional components would be benificial. Hopefully part of this PR can just be opening some discussion.

I'm also hoping this can either sit (sit closed if that's best) as a reference to be picked up by someone else, or others can test and i'm happy to re-work this code based on their feedback. Or I can try to pick back up when i have testing capibilities.

### CODE REVIEW

File: src/renderer/App.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ import { darkTheme } from "../styles/darkTheme";
 import usb from "usb";
 import { withSnackbar } from "notistack";
 
-import KeyboardSelect from "./screens/KeyboardSelect";
+import { KeyboardSelect } from "./screens/KeyboardSelect/index";
```
Comment (pm0u) [inline src/renderer/App.js:38]: The `index` distinguishes it from the existing KEyboardSelect component currently

**Comment by obra** [comment]
Paul, 

Can you drop me a note at jesse@keyboard.io?

File: src/renderer/hooks/useKeyboards.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
+// -*- mode: js-jsx -*-
+//@ts-check
```
[the remark this answers is not in the document]
Comment (pm0u) [reply src/renderer/hooks/useKeyboards.js:2]: @algernon this is for VSCode to do typechecking within this file. There are ways to have VSCode (and other IDEs) check types across _all_ files, but this allows for an incremental introduction. Also, not finalized, and open to other means of introducing some type checking

[the remark this answers is not in the document]
Comment (pm0u) [reply src/renderer/hooks/useKeyboards.js:2]: This was mostly me trying to catch as many errors as possible while working without testing ability and was the fastest way to add in some typechecking

**Comment by pm0u** [comment]
@algernon Thanks for taking a look (and taking the time to try it out!)

>Forgive my possibly silly question, I'm not very familiar with JS best practices, so to me, it would seem more practical to not have a separate ./screens/KeyboardSelect/KeyboardSelect.js, but roll it into index.js instead, and keep this import as-is.
> ...
First, this is not totally finalized in terms of structure and so this can definitely be changed. I was mainly trying to just get things moving without getting in the way of the existing KeyboardSelect (and leaving it intact for reference).

```
// this will reference the current component
import KeyboardSelect from './screens/KeyboardSelect' ;
// this will reference the one i created
import { KeyboardSelect } from './screens/KeyboardSelect/index';
```
_if_ the current keyboardselect component did not exist (the file `./screens/KeyboardSelect.js` was not present) this would work:

```
import { KeyboardSelect } from './screens/KeyboardSelect';
```

Which i think keeps the goal of "keeping the import as is" (however, see note below re: brackets)

> That does mean that we need to export a default from there, but we only export KeyboardSelect from there anyway...

i did not use a default export, but a named one.... I have worked on JS projects that do either one across the board, and i'm not sure there's an agreed "best practice." My preference is named exports, for a lot of the reasons cited [here](https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/)

Mainly, if this is the `KeyboardSelect` component, it is really useful to have it be forcibly named that everywhere for clarity (and enforcing code consistency) so you don't end up with things like

```
import theKeyboardScreen from '../../../screens/KeyboardSelect';
// other file
import KeyScreen from '../../../../../screens/KeyboardSelect';
// yet another file
import keyboardSelect from '../screens/KeyboardSelect';
```
and if you really must, you can always override the name with:

```
import { KeyboardSelect as ABetterName } from './screens/KeyboardSelect/index';
```
to be honest, i don't think that is super likely to happen with this particular component, but possibly with other smaller ones such as buttons, selects, etc as things get more split up. And consistency is nice if it makes sense for those, to do it across all.

Definitely open to discussion, this PR is more a pile of habits than choices at this point :smile:

**Comment by pm0u** [comment]
And, i should be able to do some testing of my own soon.. so hopefully that will get the change(s) that come of this past a lot of the more basic errors (such as the one you caught)

**Comment by algernon** [comment]
Getting there, getting there!

I different issues currently:

- The `Scan keyboards` button has green text.
- Disconnecting a keyboard correctly removes it from the dropdown, but reconnecting doesn't do an auto-detect anymore. 
- After reconnecting a keyboard, `Scan Keyboards` doesn't appear to do anything.

(As before, these are just high level observations, I have not looked at the code yet to figure out why these happen.)

**Comment by pm0u** [comment]
@algernon I believe this was all related to not properly resetting the state of "having found devices" on the most recent scan which disables the scan devices button for a short time. I do not have multiple boards to test, but all the issues you caught should be fixed in this latest update. Thanks for checking so thoroughly. I think there is still a little more cleanup to be done, but I believe it should be fully functional now.

**Comment by obra** [comment]
@pm0u, @algernon - what's the state of this one?

**Comment by pm0u** [comment]
to my knowledge, @obra  the state of this is that it's likely stale (although it apparently doesn't have merge conflicts). That aside, I still think from a UI perspective working towards separating things would be best, rather than having tons of logic contained in large class components. How to achieve that at this point would require another big look.

**Comment by obra** [comment]
#773 is the beginning to a new attempt to refactor KeyboardSelect in this direction. The top level component isn't yet there, but we'll get there.

**Comment by algernon** [comment]
I believe @obra's [jesse-wip](https://github.com/keyboardio/Chrysalis/tree/jesse-wip) branch has this - and a whole lot more - covered, so I'm closing this out, as further development will be on that branch.

Thank you for kick-starting the move towards hooks & functional components!

**Pull request closed without merging**